### ISSUE TITLE
RetinaNet Loss not improving

### ISSUE THREAD

**Issue by MjdMahasneh** [fizyr/keras-retinanet#1476]
I am training RetinaNet on my own dataset, for some reason the Loss (total loss) is stuck on ~1.75.  At the first epoch, Loss was around ~3.0.  So the loss did decrease, but not by much! I started encountering a loss of ~1.75 after only a few epochs (after about 25 epochs), and it consisted up until >2000 epochs!

**I am not sure what questions do I need to ask myself:**
1- Is it the implementation? if so, what do I need to look further into?
2- Is it the set up (CUDA, CuDNN, TF, Keras versions).
3- Is it any preprocessing assumptions (i.e. does the data need to be processed in a certain way?). e.g. would normalizing image value between 0-1 cause such behavior?)
4- Is it my training parameters?
5- Or maybe its something else that am messing? 

**I would appreciate any assistance on this matter.**

**-Further details :**
   -My _machine_ runs (Windows 10 64) keras 2.4, tensorflow=2.3.0, CuDNN 7.6.5, CUDA 10.1.
   -_training call_ : 
     py -3.6 keras_retinanet/bin/train.py --batch-size 5 --random-transform --steps 19 --epochs 3000 csv annotations/train_CaII.csv classes.csv
   -my _classes.csv_ contains: "REGION, 0"
   -Here is a chuck of the _report_ being printed to me:

> Epoch 2146/3000
>  2/19 [==>...........................] - ETA: 19s - loss: 2.1013 - regression_loss: 1.8193 - classification_loss: 0.2820 3/19 [===>..........................] - ETA: 23s - loss: 1.8619 - regression_loss: 1.6261 - classification_loss: 0.2357 4/19 [=====>........................] - ETA: 24s - loss: 1.8904 - regression_loss: 1.6520 - classification_loss: 0.2384 5/19 [======>.......................] - ETA: 23s - loss: 1.8843 - regression_loss: 1.6486 - classification_loss: 0.2356 6/19 [========>.....................] - ETA: 22s - loss: 1.8281 - regression_loss: 1.6010 - classification_loss: 0.2270 7/19 [==========>...................] - ETA: 21s - loss: 1.8297 - regression_loss: 1.6040 - classification_loss: 0.2257 8/19 [===========>..................] - ETA: 19s - loss: 1.8118 - regression_loss: 1.5900 - classification_loss: 0.2218 9/19 [=============>................] - ETA: 17s - loss: 1.8203 - regression_loss: 1.5914 - classification_loss: 0.228910/19 [==============>...............] - ETA: 16s - loss: 1.7932 - regression_loss: 1.5710 - classification_loss: 0.222111/19 [================>.............] - ETA: 14s - loss: 1.7679 - regression_loss: 1.5491 - classification_loss: 0.218912/19 [=================>............] - ETA: 12s - loss: 1.7828 - regression_loss: 1.5592 - classification_loss: 0.223613/19 [===================>..........] - ETA: 10s - loss: 1.7951 - regression_loss: 1.5686 - classification_loss: 0.226614/19 [=====================>........] - ETA: 9s - loss: 1.7707 - regression_loss: 1.5490 - classification_loss: 0.2217 19/19 [==============================] - 36s 2s/step - loss: 1.7543 - regression_loss: 1.5376 - classification_loss: 0.2167
> Epoch 2147/3000
>  2/19 [==>...........................] - ETA: 15s - loss: 1.6542 - regression_loss: 1.4336 - classification_loss: 0.2206 3/19 [===>..........................] - ETA: 20s - loss: 1.7254 - regression_loss: 1.5005 - classification_loss: 0.2249 4/19 [=====>........................] - ETA: 21s - loss: 1.7710 - regression_loss: 1.5461 - classification_loss: 0.2249 5/19 [======>.......................] - ETA: 20s - loss: 1.6988 - regression_loss: 1.4899 - classification_loss: 0.2089 6/19 [========>.....................] - ETA: 20s - loss: 1.7740 - regression_loss: 1.5505 - classification_loss: 0.2234 7/19 [==========>...................] - ETA: 19s - loss: 1.8155 - regression_loss: 1.5840 - classification_loss: 0.2315 8/19 [===========>..................] - ETA: 18s - loss: 1.7458 - regression_loss: 1.5261 - classification_loss: 0.2196 9/19 [=============>................] - ETA: 16s - loss: 1.7688 - regression_loss: 1.5451 - classification_loss: 0.223610/19 [==============>...............] - ETA: 15s - loss: 1.7662 - regression_loss: 1.5435 - classification_loss: 0.222711/19 [================>.............] - ETA: 13s - loss: 1.7134 - regression_loss: 1.4988 - classification_loss: 0.214612/19 [=================>............] - ETA: 12s - loss: 1.6990 - regression_loss: 1.4880 - classification_loss: 0.211113/19 [===================>..........] - ETA: 10s - loss: 1.6825 - regression_loss: 1.4743 - classification_loss: 0.208214/19 [=====================>........] - ETA: 8s - loss: 1.6989 - regression_loss: 1.4876 - classification_loss: 0.2114 19/19 [==============================] - 35s 2s/step - loss: 1.6792 - regression_loss: 1.4735 - classification_loss: 0.2057
> Epoch 2148/3000
>  2/19 [==>...........................] - ETA: 15s - loss: 1.5130 - regression_loss: 1.3538 - classification_loss: 0.1592 3/19 [===>..........................] - ETA: 19s - loss: 1.5314 - regression_loss: 1.3687 - classification_loss: 0.1627 4/19 [=====>........................] - ETA: 20s - loss: 1.6181 - regression_loss: 1.4299 - classification_loss: 0.1882 5/19 [======>.......................] - ETA: 20s - loss: 1.6536 - regression_loss: 1.4548 - classification_loss: 0.1989 6/19 [========>.....................] - ETA: 20s - loss: 1.5943 - regression_loss: 1.4040 - classification_loss: 0.1902 7/19 [==========>...................] - ETA: 19s - loss: 1.6227 - regression_loss: 1.4250 - classification_loss: 0.1976 8/19 [===========>..................] - ETA: 18s - loss: 1.6894 - regression_loss: 1.4793 - classification_loss: 0.2101 9/19 [=============>................] - ETA: 16s - loss: 1.6804 - regression_loss: 1.4710 - classification_loss: 0.209410/19 [==============>...............] - ETA: 15s - loss: 1.7283 - regression_loss: 1.5124 - classification_loss: 0.215911/19 [================>.............] - ETA: 13s - loss: 1.7554 - regression_loss: 1.5350 - classification_loss: 0.220512/19 [=================>............] - ETA: 12s - loss: 1.7404 - regression_loss: 1.5241 - classification_loss: 0.216313/19 [===================>..........] - ETA: 10s - loss: 1.7288 - regression_loss: 1.5152 - classification_loss: 0.213714/19 [=====================>........] - ETA: 8s - loss: 1.7287 - regression_loss: 1.5158 - classification_loss: 0.2129 19/19 [==============================] - 34s 2s/step - loss: 1.7656 - regression_loss: 1.5471 - classification_loss: 0.2185
> Epoch 2149/3000
>  2/19 [==>...........................] - ETA: 15s - loss: 1.4490 - regression_loss: 1.2923 - classification_loss: 0.1568 3/19 [===>..........................] - ETA: 20s - loss: 1.6261 - regression_loss: 1.4415 - classification_loss: 0.1846 4/19 [=====>........................] - ETA: 21s - loss: 1.6117 - regression_loss: 1.4294 - classification_loss: 0.1823 5/19 [======>.......................] - ETA: 21s - loss: 1.6825 - regression_loss: 1.4877 - classification_loss: 0.1949 6/19 [========>.....................] - ETA: 20s - loss: 1.7519 - regression_loss: 1.5408 - classification_loss: 0.2111 7/19 [==========>...................] - ETA: 19s - loss: 1.7471 - regression_loss: 1.5337 - classification_loss: 0.2135 8/19 [===========>..................] - ETA: 18s - loss: 1.7853 - regression_loss: 1.5643 - classification_loss: 0.2209 9/19 [=============>................] - ETA: 17s - loss: 1.7836 - regression_loss: 1.5653 - classification_loss: 0.218410/19 [==============>...............] - ETA: 15s - loss: 1.7741 - regression_loss: 1.5561 - classification_loss: 0.218011/19 [================>.............] - ETA: 13s - loss: 1.7572 - regression_loss: 1.5421 - classification_loss: 0.215112/19 [=================>............] - ETA: 12s - loss: 1.7557 - regression_loss: 1.5419 - classification_loss: 0.213813/19 [===================>..........] - ETA: 10s - loss: 1.7521 - regression_loss: 1.5400 - classification_loss: 0.212114/19 [=====================>........] - ETA: 8s - loss: 1.7726 - regression_loss: 1.5558 - classification_loss: 0.2168 19/19 [==============================] - 35s 2s/step - loss: 1.7920 - regression_loss: 1.5727 - classification_loss: 0.2193
> Epoch 2150/3000
>  2/19 [==>...........................] - ETA: 16s - loss: 1.6713 - regression_loss: 1.4906 - classification_loss: 0.1807 3/19 [===>..........................] - ETA: 20s - loss: 1.9245 - regression_loss: 1.6914 - classification_loss: 0.2331 4/19 [=====>........................] - ETA: 21s - loss: 1.9601 - regression_loss: 1.7151 - classification_loss: 0.2450 5/19 [======>.......................] - ETA: 21s - loss: 1.9395 - regression_loss: 1.6972 - classification_loss: 0.2423 6/19 [========>.....................] - ETA: 20s - loss: 1.8024 - regression_loss: 1.5767 - classification_loss: 0.2257 7/19 [==========>...................] - ETA: 19s - loss: 1.8475 - regression_loss: 1.6132 - classification_loss: 0.2343 8/19 [===========>..................] - ETA: 18s - loss: 1.8199 - regression_loss: 1.5923 - classification_loss: 0.2276 9/19 [=============>................] - ETA: 16s - loss: 1.7762 - regression_loss: 1.5551 - classification_loss: 0.221110/19 [==============>...............] - ETA: 15s - loss: 1.7958 - regression_loss: 1.5709 - classification_loss: 0.224811/19 [================>.............] - ETA: 13s - loss: 1.7749 - regression_loss: 1.5565 - classification_loss: 0.218512/19 [=================>............] - ETA: 12s - loss: 1.7963 - regression_loss: 1.5741 - classification_loss: 0.222213/19 [===================>..........] - ETA: 10s - loss: 1.7884 - regression_loss: 1.5663 - classification_loss: 0.222114/19 [=====================>........] - ETA: 8s - loss: 1.7942 - regression_loss: 1.5715 - classification_loss: 0.2227 19/19 [==============================] - ETA: 0s - loss: 1.7985 - regression_loss: 1.5720 - classification_loss: 0.2264
> Epoch 02150: saving model to ./snapshots\resnet50_csv_2150.h5
> 19/19 [==============================] - 36s 2s/step - loss: 1.7985 - regression_loss: 1.5720 - classification_loss: 0.2264
> Epoch 2151/3000
>  2/19 [==>...........................] - ETA: 17s - loss: 1.4341 - regression_loss: 1.2471 - classification_loss: 0.1870 3/19 [===>..........................] - ETA: 21s - loss: 1.5576 - regression_loss: 1.3568 - classification_loss: 0.2008 4/19 [=====>........................] - ETA: 22s - loss: 1.5970 - regression_loss: 1.3972 - classification_loss: 0.1999 5/19 [======>.......................] - ETA: 24s - loss: 1.6943 - regression_loss: 1.4764 - classification_loss: 0.2179 6/19 [========>.....................] - ETA: 22s - loss: 1.6899 - regression_loss: 1.4770 - classification_loss: 0.2130 7/19 [==========>...................] - ETA: 21s - loss: 1.7079 - regression_loss: 1.4939 - classification_loss: 0.2141 8/19 [===========>..................] - ETA: 19s - loss: 1.6629 - regression_loss: 1.4579 - classification_loss: 0.2050 9/19 [=============>................] - ETA: 18s - loss: 1.6838 - regression_loss: 1.4703 - classification_loss: 0.213510/19 [==============>...............] - ETA: 16s - loss: 1.7189 - regression_loss: 1.5005 - classification_loss: 0.218411/19 [================>.............] - ETA: 14s - loss: 1.7258 - regression_loss: 1.5083 - classification_loss: 0.217512/19 [=================>............] - ETA: 12s - loss: 1.7386 - regression_loss: 1.5214 - classification_loss: 0.217213/19 [===================>..........] - ETA: 11s - loss: 1.7828 - regression_loss: 1.5577 - classification_loss: 0.225114/19 [=====================>........] - ETA: 9s - loss: 1.7639 - regression_loss: 1.5424 - classification_loss: 0.2215 19/19 [==============================] - 36s 2s/step - loss: 1.8384 - regression_loss: 1.6044 - classification_loss: 0.2340
> Epoch 2152/3000
>  2/19 [==>...........................] - ETA: 16s - loss: 1.8173 - regression_loss: 1.5916 - classification_loss: 0.2257 3/19 [===>..........................] - ETA: 20s - loss: 1.8859 - regression_loss: 1.6520 - classification_loss: 0.2339 4/19 [=====>........................] - ETA: 21s - loss: 1.8495 - regression_loss: 1.6182 - classification_loss: 0.2313 5/19 [======>.......................] - ETA: 21s - loss: 1.8532 - regression_loss: 1.6228 - classification_loss: 0.2304 6/19 [========>.....................] - ETA: 20s - loss: 1.8635 - regression_loss: 1.6229 - classification_loss: 0.2406 7/19 [==========>...................] - ETA: 19s - loss: 1.9119 - regression_loss: 1.6658 - classification_loss: 0.2461 8/19 [===========>..................] - ETA: 18s - loss: 1.8691 - regression_loss: 1.6329 - classification_loss: 0.2363 9/19 [=============>................] - ETA: 16s - loss: 1.8461 - regression_loss: 1.6120 - classification_loss: 0.234210/19 [==============>...............] - ETA: 15s - loss: 1.8417 - regression_loss: 1.6074 - classification_loss: 0.234311/19 [================>.............] - ETA: 13s - loss: 1.8497 - regression_loss: 1.6137 - classification_loss: 0.236012/19 [=================>............] - ETA: 12s - loss: 1.7885 - regression_loss: 1.5619 - classification_loss: 0.226613/19 [===================>..........] - ETA: 10s - loss: 1.7890 - regression_loss: 1.5638 - classification_loss: 0.225114/19 [=====================>........] - ETA: 8s - loss: 1.8038 - regression_loss: 1.5745 - classification_loss: 0.2292 19/19 [==============================] - 34s 2s/step - loss: 1.7564 - regression_loss: 1.5372 - classification_loss: 0.2192
> .
> .
> .
> .

**Comment by Karthik-U-94** [comment]
Same Issue I am using my own data, I have tried for 300 Epochs Loss is not decreasing, It is fluctuating around 1.07 to 1.05.
Env Details:

**Comment by MjdMahasneh** [comment]
@Karthik-U-94 Could you please give me more context, what type of data are you using (e.g. RGB, monochrome..etc)? is your data pre-processed in anyway and is there any normalization? 

Maybe we can eliminate possibilities by finding similarities in our settings!

**Comment by Karthik-U-94** [comment]
@MjdMahasneh 
Data is having ~98% of Grayscale images and only 2% RGB. 
for 1 image will output 32 images including with Tranformed Points
Data is custom Augmented:
                                             1) Rotation: -90,-45,45,90 (4 images)
                                             2) Flip (2 images)
                                             3) Affline Scale: 0.5 and 1.5 (1 image)
                                             4) Noise: Salt-and-pepper (8 images) 
                                             
Resize (600x420) (All 16 images)

Image size: 1) 2384x1684px (16 images)
                    2) 600x420px (16 images)

So, 50% of my data is in 2384x1684px and 50% is in 600x420px 

Pretrainined model used from : ` https://github.com/fizyr/keras-retinanet/releases/download/0.5.1/resnet101_oid_v1.0.0.h5`
Backbone defult is changed to `Retinanet101`
training call: `!keras-retinanet/keras_retinanet/bin/train.py --freeze-backbone --random-transform --weights {PRETRAINED_MODEL} --batch-size 8 --steps 297 --epochs 300 csv {ANNOTATIONS_FILE} {CLASSES_FILE}`
